docs/proposals/20191016-e2e-test-framework.md

```diff
@@ -0,0 +1,199 @@
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Behaviors to test](#behaviors-to-test)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Alternatives](#alternatives)
+- [Upgrade Strategy](#upgrade-strategy)
+- [Additional Details](#additional-details)
+  - [Test Plan](#test-plan)
+  - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+- [Implementation History](#implementation-history)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+If this proposal adds new terms, or defines some, make the changes to the book's glossary when in PR stage.
+
+Basic e2es:
+The Reconcilers are exercised by creating a cluster and deleting a cluster
+Comprehensive e2es:
+Many of the components and behaviors of Cluster API are exercised.
+Creates a number of clusters and tears them down.
+
+## Summary
+
+Cluster API's providers could benefit from having a set of generic behavioral e2e tests. 
+
+| infra provider | has basic e2es | has comprehensive e2e suite |
+| --- |:---:|:---:|
+| CAPD | ❌ | ❌ |
+| CAPA | ✅ | ❌ |
+| CAPV | ❌ | ❌ |
+| CAPG | ✅ | ❌ |
+
+| bootstrap provider | has basic e2es | has comprehensive e2e suite |
+| --- |:---:|:---:|
+| CABPK| ❌ | ❌ |
+
+A pluggable set of behavioral tests for Cluster API Providers would prove beneficial for the project health as a whole.
+
+As stated in the non-goals, this proposal does not intend to define the conformance requirements for Cluster API providers.
+
+## Motivation
+
+Every infrastructure and bootstrap provider wants to know if it “works”, but testing today is a manual process. Developers test one or two well-trod scenarios if testing at all before merging PRs. This test suite, when properly configured, would provide a baseline signal for any given provider for more scenarios than any single developer would reasonably spend time testing.
+
+A bonus to creating a pluggable suite of behaviors is the consolidation of expectations in one place. The test suite is a direct reflection of the core provider and thus is versioned along with the cluster-api repository.
+
+### Goals
+
+* Define sets of tests that all providers can pass
+  * End-to-end tests
+  * (optional) Unit tests
+* Build a test suite as a go library that providers can import, supply their own types/objects that satisfy whatever interface is required of the tests
+  * Test suite can be run in parallel
+  * Tests should be organized so providers can specify certain tests to run or skip. For example, a fast-focus could be a PR blocking job whereas the slow tests could run periodically
+* Create documentation for plugging a provider into this library
+    * Particularly tricky bits here will be providing guidance on how to manage secrets
```

Comment:
Are we planning to rely on controller-runtime everywhere for this framework? As in, client, envtest, etc.

Reply:
I like the idea of using the controller-runtime client, but I think we'd probably be better served in e2e testing by using `kind` rather than `envtest`